File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private int gcd(int a, int b) {
        if(b == 0) return a;
        return gcd(b, a % b);
    }

    public ListNode insertGreatestCommonDivisors(ListNode head) {
        ListNode curr = head;
        while(curr.next != null) {
            int val = gcd(curr.val, curr.next.val);
            ListNode temp = curr.next, newNode = new ListNode(val);
            curr.next = newNode;
            newNode.next = temp;
            curr = temp;
        }

        return head;
    }
}

File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def gcd(self, a, b):
        if b == 0: 
            return a
        return self.gcd(b, a % b)

    def insertGreatestCommonDivisors(self, head):
        curr = head
        while curr.next:
            val = self.gcd(curr.val, curr.next.val)
            temp, new_node = curr.next, ListNode(val)
            curr.next = new_node
            new_node.next = temp
            curr = temp
        
        return head
        